Fix text addition / removal bug (#28)

wallstop opened the pull request:
# Overview
See #28 for the issue

# The Fix (Fixes #28, Fixes #5)
I've done five things here:
1. Cached a reference to `OnTextChanged` such that we're not re-creating new delegates on every OnEnable/OnDisable call
2. Keep stateful knowledge of whether this delegate is attached
3. Update `OnTextChanged` to not call `UpdateMeshes`. This allows for safe addition, removal, and automatic re-calculation of the meshes
4. Changed ApplyModifiers to not throw. This is important - TweenManager may be initialized early than the text meshes. The hook into the TEXT_CHANGED_EVENT will be triggered when the text meshes are safely initialized.
5. Changed ForceMeshUpdate to only happen in non-PlayMode. I was having issues related to this in Play mode - I think if it's in Play Mode, things will initialize naturally.

Now, I edit text while the editor isn't playing and have it behave expectedly, AND I can do the same in play mode.

AlicanHasirci commented:
This works great with single texts but when there are multiple texts, since the unchanged text mesh is not regenerated, the changes get reapplied.

### Reproduction
1.  Add multiple texts to TweenManger
2. Set progress above 0 so there are changes to apply
3. Start changing one of the texts

wallstop commented:
> This works great with single texts but when there are multiple texts, since the unchanged text mesh is not regenerated, the changes get reapplied.
> ### Reproduction
> 
>     1. Add multiple texts to TweenManger
> 
>     2. Set progress above 0 so there are changes to apply
> 
>     3. Start changing one of the texts

Great catch and thanks for the repro. I believe I've fixed this as well as some related bugs due to text changing.

Here's what I've done:
1. Added some stateful knowledge about text -> last known vertex count. When handling multiple texts within a single TweenManager, we still want to restore the meshes of texts that are *not* the current one. However, since they're all stored in a giant array and we don't know what our "current" value of allocated space within that array is (due to all knowledge of "size" being current state, not state when the array was computed), this was needed. The concept is pretty simple - on init, cache text -> vertex count. On Dispose, clear the cache. On text change, if the text being changed is in the cache, do not update its mesh, but advance the pointer in our buffer by the appropriate (cached, not current) value.
2. Fixed a bug related to visible character count. To test this I created several texts and added/removed characters from them. When adding characters, everything worked as expected. However, when removing characters, I found out that the current TMP impl does some optimizations related to array resizing. That is, it doesn't do anything to character data in the `text.textInfo.characterInfo` array - it keeps the old state there, *including* the `isVisible` property (!!!). What it *does* do, though, is expose a `characterCount` property, which appears to accurately reflect the state of "which character data is accurate". I had to update this in two places - the GetVisibleCharCount method as well as in CreateCharDataArray that calls it
3. Fixed a bug where I forgot to set _eventAdded=true when we add the event, resulting in a listener leak.

AlicanHasirci commented:
> > This works great with single texts but when there are multiple texts, since the unchanged text mesh is not regenerated, the changes get reapplied.
> > ### Reproduction
> > ```
> > 1. Add multiple texts to TweenManger
> > 
> > 2. Set progress above 0 so there are changes to apply
> > 
> > 3. Start changing one of the texts
> > ```
> 
> Great catch and thanks for the repro. I believe I've fixed this as well as some related bugs due to text changing.
> 
> Here's what I've done:
> 
> 1. Added some stateful knowledge about text -> last known vertex count. When handling multiple texts within a single TweenManager, we still want to restore the meshes of texts that are _not_ the current one. However, since they're all stored in a giant array and we don't know what our "current" value of allocated space within that array is (due to all knowledge of "size" being current state, not state when the array was computed), this was needed. The concept is pretty simple - on init, cache text -> vertex count. On Dispose, clear the cache. On text change, if the text being changed is in the cache, do not update its mesh, but advance the pointer in our buffer by the appropriate (cached, not current) value.
> 2. Fixed a bug related to visible character count. To test this I created several texts and added/removed characters from them. When adding characters, everything worked as expected. However, when removing characters, things were not working as expected. I found out that the current TMP impl does some optimizations related to array resizing. That is, on character removal, it doesn't do anything to character data in the `text.textInfo.characterInfo` array - it keeps the old state there, _including_ the `isVisible` property (!!!). What it _does_ do, though, is expose a `characterCount` property, which appears to accurately reflect the state of "which character data is accurate". I had to update this in two places - the GetVisibleCharCount method as well as in CreateCharDataArray that calls it
> 3. Fixed a bug where I forgot to set _eventAdded=true when we add the event, resulting in a listener leak.

Great work and thanks for elaborating on the changes, also a great catch with  `text.textInfo.characterInfo` not changing 🙏 
Seems like every problem stemming from text related changes, points to code smell coming from my hasty multi-text support commit. I'll be working on a solution to fix this issues in general and move the responsibility of individual texts from TweenManager to a different class.
